### PR TITLE
Optimize DB Upserts & Bulk Inserts

### PR DESCRIPTION
**Pull request by harry1453**

### CODE REVIEW

File: src/main/kotlin/brs/db/sql/SqlDb.kt [hunk 1]
```diff
@@ -41,6 +41,7 @@ internal class SqlDb(private val dp: DependencyProvider) : Db {
         return if (con == null) {
             DSL.using(cp, dialect, settings)
         } else {
+            // Using transaction connection. Use static statements. TODO why?
```
Comment (llybin) [inline src/main/kotlin/brs/db/sql/SqlDb.kt]: https://github.com/burst-apps-team/burstcoin/commit/aa5b04a66fe55a39eb05ad3cdf5759669b3ba111#diff-d5cd48c303de3c01c50833e2442f62abR231

File: src/main/kotlin/brs/util/db/Upsert.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+fun DSLContext.upsert(table: Table<*>, columns: Collection<Field<*>>, keys: Collection<Field<*>>, values: Collection<Array<*>>): Batch {
+    require(columns.isNotEmpty() && keys.isNotEmpty() && values.isNotEmpty()) { "Columns, keys and values must not be empty" }
+    return if (dialect().supportsMerge()) {
+        val query = batch(upsert(table, keys, mapOf(columns, NULL_ITERABLE)))
+        values.forEach { value ->
+            query.bind(*value)
+        }
+        query
+    } else {
+        // TODO I can't get the bind values to work both in the insert clause and the update clause for SQLite.
```
Comment (llybin) [inline src/main/kotlin/brs/util/db/Upsert.kt]: Probably, the initial problem here https://github.com/burst-apps-team/burstcoin/pull/375/files#diff-d83b87a3832813cb3d86fbca3c3d0b78R67 and not in sqlite? We haven't nullable fields, if no column/value - need set default value for this column, not null.

File: src/main/kotlin/brs/db/sql/SqlDb.kt (hunk 1 above)
Comment (harry1453) [reply src/main/kotlin/brs/db/sql/SqlDb.kt]: Yeah, I don't remember why I did this..

File: src/main/kotlin/brs/util/db/Upsert.kt (hunk 2 above)
Comment (harry1453) [reply src/main/kotlin/brs/util/db/Upsert.kt]: But the null fields are placeholders for bind values. The other DBs accept it ok. Maybe we need named bind variables..

File: src/main/kotlin/brs/db/sql/SqlSubscriptionStore.kt [hunk 3]
```diff
@@ -19,28 +18,69 @@ internal class SqlSubscriptionStore(private val dp: DependencyProvider) : Subscr
...
+                    SUBSCRIPTION.ID,
+                    SUBSCRIPTION.SENDER_ID,
+                    SUBSCRIPTION.RECIPIENT_ID,
+                    SUBSCRIPTION.AMOUNT,
+                    SUBSCRIPTION.FREQUENCY,
+                    SUBSCRIPTION.TIME_NEXT,
+                    SUBSCRIPTION.HEIGHT,
+                    SUBSCRIPTION.LATEST
+                )
+                private val upsertKeys = upsertColumns // TODO do we really need every column unique? what's the point?
```
Comment (llybin) [inline src/main/kotlin/brs/db/sql/SqlSubscriptionStore.kt]: Could be more than one record SENDER_ID-RECIPIENT_ID on one HEIGHT? - I think SENDER_ID-RECIPIENT_ID-HEIGHT - is unique.

Comment (llybin) [reply src/main/kotlin/brs/db/sql/SqlSubscriptionStore.kt]: subscriptions should be never updated, only inserts, iiiif get rid upsert .... xD

Comment (llybin) [reply src/main/kotlin/brs/db/sql/SqlSubscriptionStore.kt]: if see on code above, there unique on transaction_id-height...need to think. Maybe enough only transaction_id (which id)? Transaction_id is always unique per table? Can we ident row by transaction_id? I think yes but not sure.

**Comment by shefass** [comment]
I not looked to your stuff.. but now assetID == AssetIsuanceTransactionID,
AT_ID == ATCreationTransactionID...so I guess same with subscriptions.
So Transaction in which you create asset or AT tied to it. Probably you
know this stuff

2020-03-29, sk, 18:27 Lev Lybin <notifications@github.com> rašė:

**Comment by llybin** [comment]
tests are passed, but I get errors on sync

**Comment by harry1453** [comment]
@shefass That is correct.

@llybin hmm, I haven't tested on H2 yet, thanks for the heads up, will try it. This pull is not ready yet anyway.

**Pull request merged**